test/helpers/restricted-nori.ts

```diff
@@ -33,16 +30,14 @@ export const restrictRemovalProceeds = async ({
   removalAmountsToRestrict: number[];
 }): Promise<void> => {
   const { rNori, bpNori } = testSetup;
-  await Promise.all(
-    removalIds.map((id, index) => {
-      const userData = formatTokensReceivedUserData(id);
-      return bpNori.send(
-        rNori.address,
-        removalAmountsToRestrict[index],
-        userData
-      );
-    })
-  );
+  const promiseArray = [];
+  for (const [index, id] of removalIds.entries()) {
+    promiseArray.push(
+      rNori.mint(removalAmountsToRestrict[index], id),
+      bpNori.transfer(rNori.address, removalAmountsToRestrict[index])
```

Comment:
I don't think this is doing what you think. Just because `await` is omitted doesnt mean the function wasn't called. As such, the await after the loop could end up with an array of pending/fulfilled promises.

Reply:
isn't this basically the same thing that's happening when you use .map and generate an array of promises?  it's just that extra promises are being added to the array? the functions are being called but nothing is being awaited serially, and then it's just a typical `Promise.all` on all those promises.

if I still have this wrong, do you know the way to do this better?

Reply:
```ts
params = [3000, 3000, 3000];
demoAsyncForLoop = async () => {
  promises = [];
  i = 0;
  for (const x of params) {
    promises.push(timeout(x, i));
    i++;
  }
  await Promise.all(promises);
};
demoPromiseAll = async () => {
  await Promise.all(
    params.map((p, index) => {
      return timeout(p, index);
    })
  );
};
```
![image](https://user-images.githubusercontent.com/18407013/176508977-dd0f07e3-f7e0-4710-ba83-7e7f1831e052.png)

Reply:
I see what you're trying to say now. The problem is that this is still not following a great pattern though (for loops + declaring arrays and pushing into them is imperative and non-functional and should be avoided for their side effects that they introduce).  [[more](https://github.com/MostlyAdequate/mostly-adequate-guide/blob/822af0ce4a15b64ea6470f6dee02cf6fffd39793/ch06.md#declarative-coding)]

IMO, It should just be something like this

```ts
 await Promise.all(
    removalIds.map(async (id, index) => {
      return Promise.all([
        rNori.mint(removalAmountsToRestrict[index], id),
        bpNori.transfer(rNori.address, removalAmountsToRestrict[index]),
      ]);
    })
  );
```

Reply:
[another good resource](https://github.com/you-dont-need/You-Dont-Need-Loops)

Reply:
ahh, I see.  I agree that when I wrote this something about it felt wrong, even if it's technically doing what I want it to.  I see how this also works and doesn't unnecessarily serialize anything. updated.

Reply:
Thanks for seeing this one through! 💯